### PR TITLE
Experimental support for ti.func multi-return with scalar values #536 (3)

### PR DESCRIPTION
**Pull request by archibate**
Related issue id = #536

### CODE REVIEW

**Comment by archibate** [comment]
Sorry for messing `i538` here... but this multi-return couldn't be done without that AST fix :)

**Comment by yuanming-hu** [comment]
Thanks! This is really helpful. Sorry about the forced-push. I had to maintain a clear changeset for reviewing.

**Comment by yuanming-hu** [comment]
Would it be possible to briefly test this? For example, `max` and `abs`. Note that we should not spend too much time on this - The IR should be slightly improved to systematically support functions, including those with multiple returns. I'm still thinking about how that should be done.

**Comment by archibate** [comment]
This is breaking test because ti.Matrix return is not supported. Should `__retval = ti.expr_init(mat.id())`

**Comment by archibate** [comment]
How to zero-initialize given types? eg. `ti.expr_zero_init(ti.f32)`

**Comment by archibate** [comment]
Suggestions are super welcome.

**Comment by yuanming-hu** [comment]
Sorry I was on a meeting this morning. Can we try something like `ti.expr_init(Matrix.zero(2, 2, dt=ti.f32))`?

**Comment by archibate** [comment]
> Sorry I was on a meeting this morning. Can we try something like `ti.expr_init(Matrix.zero(2, 2, dt=ti.f32))`?

That's exactly what I've done. The difference is because, we can't write `def func() -> ti.Matrix(2, 2)`, since `ti.Matrix(2, 2)` will be evaluated and you get segfault. So we need `ti.mat(2, 2)` - a partially-applicated version of `ti.Matrix`.

File: python/taichi/lang/impl.py [hunk 1]
```diff
@@ -5,6 +5,18 @@
...
 
 
+def type_init(rhs):
+  import taichi as ti
+  if rhs is None:
+    return ti.expr_init(None) # alloca
+  elif isinstance(rhs, taichi_lang_core.DataType):
+    return ti.expr_init(ti.cast(0, rhs))
+  elif isinstance(rhs, ti.mat):
+    return ti.expr_init(rhs.init())
```
Comment (archibate) [inline python/taichi/lang/impl.py]: Notice here, we used `ti.mat.init()`, where `ti.mat` contains n and m infomations by `->` signature.

**Comment by archibate** [comment]
Check out this example:
```py
import taichi as ti

ti.init()#print_preprocessed=True)

x = ti.var(ti.f32, ())
u = ti.Vector(2, dt=ti.f32, shape=())
z = ti.Vector(3, dt=ti.f32, shape=())

mprint = print

@ti.kernel
def func(t: ti.f32):
  a = ti.Vector([2, 3])
  x[None] = fake_absolute(t)
  u[None] = normalized(a)
  z[None] = zero_by_default(a)

@ti.func
def fake_absolute(x): # will guess return type by alloca, better not
  if x > 0:
    return float(x)
  else:
    return int(-x)

@ti.func
def zero_by_default(x) -> ti.vec(3, dt=ti.i32):
  pass

@ti.func
def normalized(v) -> ti.vec(2, dt=ti.f32):
  return v / 2

func(233.33)
print(x[None])
func(-233.33)
print(x[None])
print(u[None][0])
print(u[None][1])
print(z[None][2])
(ast) [bate@archit taichi]$ b feature.my.py 
(ast) [bate@archit taichi]$ cat feature.my.py  
import taichi as ti

ti.init()#print_preprocessed=True)

x = ti.var(ti.f32, ())
u = ti.Vector(2, dt=ti.f32, shape=())
z = ti.Vector(3, dt=ti.f32, shape=())

@ti.kernel
def func(t: ti.f32):
  a = ti.Vector([2, 3])
  x[None] = fake_absolute(t)
  u[None] = normalized(a)
  z[None] = zero_by_default(a)

@ti.func
def fake_absolute(x): # will guess return type by alloca, better not
  if x > 0:
    return float(x)
  else:
    return int(-x)

@ti.func
def zero_by_default(x) -> ti.vec(3, dt=ti.i32):
  pass

@ti.func
def normalized(v) -> ti.vec(2, dt=ti.f32):
  return v / 2

func(233.33)
print(x[None])
func(-233.33)
print(x[None])
print(u[None][0])
print(u[None][1])
print(z[None][2])
```

**Comment by yuanming-hu** [comment]
Considering the fact that
 - We want `@ti.func` to be generic, so that, say, your `normalize` function will work not only for `Vector(2)` but also `Vector(3)`.
 - Very soon we will migrate to a system that builds functions using Taichi IR instead of Python inlining, so here we should consider this migration cost

I suggest 
 - In the `ti.func` transform, insert a `taichi_lang_core.initialize_funcion_scope` call at the very beginning and  `taichi_lang_core.finalize_funcion_scope` at the end. This helps us track the return values.
 - Then transforming return statements, insert a `taichi_lang_core.create_return`.
 - In each function_scope, on the first `create_return` call, allocate a set of mutable local variables to store the return values. If returning a scalar, allocate a scalar; if returning a matrix, allocate N mutable local variables, where N is the number of entries in the matrix. **Note that instead of using type annotation (which stops generic programming), the return type of a function is defined by the first return statement in the function body**
 - For each return, store the return value to the mutable local variables. The data type of these return values will be inferred on the **first** (in AST order) `LocalStoreStmt` during type checking.
 - At the end of the function, simply return the mutable local variables. Maybe reconstruct Python classes (e.g. Matrix) so that users get a Matrix instead of a tuple of `ti.Expr`

The switch to inline-based functions to IR-based functions will not be too hard if we follow this strategy.

**Comment by archibate** [comment]
> We want @ti.func to be generic, so that, say, your normalize function will work not only for Vector(2) but also Vector(3).

This is worked only with inline ones, right? Like `ti.template()` does.

**Comment by archibate** [comment]
Sounds cool! Will work on this trr.

**Comment by archibate** [comment]
I also have a idea that arguments and returns can be unified like...`in` and `out` keywords, eg.:
```py
def func(x: ti.out(), y: ti.out(), t): # x, y are l-values from caller
  x = ti.Vector([t, t]) # x.assign(...), calls ti.Out.assign
  y = t + 1
  t = 233 # t.assign(...), calls ti.Expr.assign, won't affect caller's value.
```
With this we can return values with different type! And the keyword `return` is treated as a special case of them...

**Comment by yuanming-hu** [comment]
> I also have a idea that arguments and returns can be unified like...in and out keywords

Yeah, this could work too. The user them have to allocate a return value outside function calling though :-(

**Comment by yuanming-hu** [comment]
> > We want @ti.func to be generic, so that, say, your normalize function will work not only for Vector(2) but also Vector(3).
> 
> This is worked only with inline ones, right? Like `ti.template()` does.

Good point. I need to think deeper into this.

**Comment by archibate** [comment]
> > I also have a idea that arguments and returns can be unified like...in and out keywords
> 
> Yeah, this could work too. The user them have to allocate a return value outside function calling though :-(

But we can transform `x = foo()` into
```py
__tmp = alloca()
foo(__tmp)
x = __tmp
```
with our `ASTTransformer`!

Also transform
```py
def func():
  return 233
```
into
```py
def func(__retval: ti.out()):
  __retval.assign(233)
```

**Comment by yuanming-hu** [comment]
I guess the issue is `__tmp = alloca()` has no matrix size information

**Comment by yuanming-hu** [comment]
> > > We want @ti.func to be generic, so that, say, your normalize function will work not only for Vector(2) but also Vector(3).
> > 
> > This is worked only with inline ones, right? Like `ti.template()` does.
> 
> Good point. I need to think deeper into this.

I guess ultimately we will end up with `ti.func` that behaves just like `ti.kernel`: instantiation is lazy, and templates are supported. We use a hash table to avoid redundant instantiations. All these happen at the Python frontend, and the Taichi IR is fully static, with functions being statically typed. Matrices/vectors are likely flattened into tuples of scalars.

**Comment by archibate** [comment]
> We use a hash table to avoid redundant instantiations.

So we use `(funcname, argtype)` as hash key. Then like C++ templates, compile for each argtype combination that used?

**Comment by yuanming-hu** [comment]
Exactly!!!

**Comment by archibate** [comment]
Trying to change `current_ast_builder()` to my `FuncBodyStmt`.

**Comment by archibate** [comment]
Return looks very like break in do-while(0), may do with inline.

**Comment by archibate** [comment]
break: jump out of current loop scope
return: jump out of current function scope
Can you tell me more details on how break is impled?

**Comment by archibate** [comment]
Will take class from trr, may not have so much time coding any more. Also will go to bed earlier.

**Comment by archibate** [comment]
Btw, trigger python error after exiting gdb
```
RuntimeError: Segmentation fault
```
Cool!

**Comment by yuanming-hu** [comment]
Thanks! Originally I thought you are going to add support for `return`'s in Taichi AST frontend, but now I realized that you are directly adding functions support to the Taichi IR. I think this is great (and more challenging). Please let me think about more details of the high-level IR design strategy so that we will end up with a clean & maintainable IR design for Taichi functions.

> Will take class from trr, may not have so much time coding any more. Also will go to bed earlier

Sounds good! Have fun at school!

**Comment by archibate** [comment]
Seeking for `builder->CreateFunctionBody` now.

**Comment by yuanming-hu** [comment]
I suggest that we first figure out an IR design strategy before implementing it...

**Comment by archibate** [comment]
Currently I see `BreakStmt` is lowered into `WhileControlStmt` which seek for `after_break` and branch to there:
https://github.com/taichi-dev/taichi/blob/c826de521d254745db556835e322dd2e0cfdbfa0/taichi/codegen/codegen_llvm.h#L709-L717
I'm going to use this method, a conditionless branch to `after_function` (1), and store return value somewhere (2).

**Comment by archibate** [comment]
Got `Basic Block in function 'func_c4_0__kernel_0_serial' does not have terminator!`.

**Comment by yuanming-hu** [comment]
> Got Basic Block in function 'func_c4_0__kernel_0_serial' does not have terminator!.

This is likely because a `return` statement is missing. Every llvm function should be terminated by a `ret`, even for `void` functions (use `builder->CreateRetVoid`).

**Comment by archibate** [comment]
Tried adding to different place, become `Terminator found in the middle of a basic block!` now.

**Comment by yuanming-hu** [comment]
> Tried adding to different place, become Terminator found in the middle of a basic block! now.

Oh, please print out the LLVM IR to see if you have a return statement at the middle of any basic block. Note that LLVM follows a control flow graph + basic block IR strategy, and return statements can only be the **last** statement of each basic block.

**Comment by archibate** [comment]
```
==========
kernel {
  $0 = offloaded  {
    function body
    return $5
    return $8
    function leave
    <i32x1>  $5 = const [666]
    print a, $5
  }
}
==========
```
(I put `CreateRetVoid` in `function leave`. The function is inlined for now)

**Comment by yuanming-hu** [comment]
Oh I see. For the LLVM backend, maybe we can simply create a real LLVM function (`llvm::Function`) and then insert an LLVM function call (`call`)?

Or maybe it's easier if you start with adding function body definition/function calls to your OpenGL backend, since you will then be issuing C++ code which is much easier to debug.

(Of course we will need a function inlining pass sooner or later)

**Comment by yuanming-hu** [comment]
After digging into your implementation, I feel like there's a misunderstanding. 

We want to end up with an IR system that supports function calls, for example
```
==========
func <i32x1> max(<i32x1> a, <i32x1> b) {
  <i32x1> cond = a > b
  if (cond) {
    return a
  } else {
    return b
  }
}

kernel {
  $0 = offloaded  {
    <i32x1> $1 = const [233]
    <i32x1> $2 = const [666]
    <i32x3> $3 = call max ($1, $2)            // this func call can also be optionally inlined. 
    print a, $3
  }
}
==========
```

`taichi_lang_core.initialize_funcion_scope` and `taichi_lang_core.finalize_funcion_scope` are just for creating the function AST, which will get lowered into a SSA function body like the `max` func above.

**Comment by yuanming-hu** [comment]
After implementing this IR extension, multi-return functions could be handled systematically. However, this may be a lot of work and probably we should use multiple PRs. I can go ahead and implement the scaffold of the IR system if you want.

(Originally the quick & hacky solution is to create a local variable in the Python scope to store the return value, without creating any Taichi functions since the Python function will get inlined. However, that's just a temporary solution.)

**Comment by archibate** [comment]
Was get rid of `Terminator found in the middle of a basic block!` impossible without creating a `llvm::Function`? I'm thinking about step-by-step strategy, and I understand what we'll get at the end.

**Comment by yuanming-hu** [comment]
> Was get rid of `Terminator found in the middle of a basic block!` impossible without creating a `llvm::Function`? I'm thinking about step-by-step strategy, and I understand what we'll get at the end.

When a function is inlined, it's return value should be store in a local variable for its caller to read, instead of creating a return. In LLVM, `return` means exiting the current function (e.g. the outer-most level parallel for).

However, ideally, we don't want to inline everything - that's why we are creating functions in Taichi IR, instead of inlining everything in the Python scope.

**Comment by archibate** [comment]
> I can go ahead and implement the scaffold of the IR system if you want.

Oh that would be great! I don't have too much time on reading LLVM docs during school days, if you could help, may helps me a lot!

**Comment by yuanming-hu** [comment]
> Oh that would be great! I don't have too much time on reading LLVM docs during school days, if you could help, may helps me a lot!

Sounds good! I may be slow since I have many things to do simultaneously, but I already have a clear plan in my mind :-)

**Comment by archibate** [comment]
> In LLVM, `return` means exiting the current function (e.g. the outer-most level parallel for).

So `CreateRetVoid` is actually `exit()` in LLVM?

> However, ideally, we don't want to inline everything - that's why we are creating functions in Taichi IR, instead of inlining everything in the Python scope.

Yeah, maybe translate `FuncBody` to inline ones is what lowering should do... just like how `for i in range(3)` is expanded.

**Comment by yuanming-hu** [comment]
> So `CreateRetVoid` is actually `exit()` in LLVM?

It's just like `return` in C++ and will exit the current function scope.

> Yeah, maybe translate `FuncBody` to inline ones is what lowering should do... just like how `for i in range(3)` is expanded.

The Python-side AST generation currently does the inlining job. You are right that the mechanism is the same as `for i in ti.static(range(3))`. (Note `ti.static` here)

I have to go to sleep as I have an early meeting tomorrow morning. Good night!

**Comment by archibate** [comment]
Have a good dream!

**Comment by archibate** [comment]
How to add `llvm::Function` other than main kernel?

**Comment by archibate** [comment]
tmp give up, since not in v0.6.0 roadmap.

**Comment by yuanming-hu** [comment]
> How to add llvm::Function other than main kernel?

> tmp give up, since not in v0.6.0 roadmap.

Let me spend a few more days further systematically thinking about this issue. Introducing new components to the IR system needs careful considerations. Sorry about the delay.

**Comment by archibate** [comment]
I found that `llvm::Function` must spec super clear return type like `Int32`, does templates solve return type too?
Btw, failed due to spdlog after merged with master.

**Comment by yuanming-hu** [comment]
> I found that llvm::Function must spec super clear return type like Int32, does templates solve return type too?

That's exactly what I'm thinking about. We need the IR to lower the compound return values into what LLVM supports. Again, designing a new IR extension needs a lot of careful considerations and I would spend some time on it before starting implementation.

**Comment by archibate** [comment]
Replaced by #612

**Pull request closed without merging**